pkg/icingaredis/telemetry/heartbeat.go

```diff
@@ -0,0 +1,115 @@
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/go-redis/redis/v8"
+	"github.com/icinga/icingadb/internal"
+	"github.com/icinga/icingadb/pkg/com"
+	"github.com/icinga/icingadb/pkg/icingaredis"
+	"github.com/icinga/icingadb/pkg/logging"
+	"github.com/icinga/icingadb/pkg/periodic"
+	"github.com/icinga/icingadb/pkg/utils"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"runtime/metrics"
+	"strconv"
+	"time"
+)
+
+// ha represents icingadb.HA to avoid import cycles.
+type ha interface {
+	State() (weResponsibleTs time.Time, weResponsible, otherResponsible bool)
+}
+
+// jsonMetricValue allows JSON-encoding a metrics.Value.
+type jsonMetricValue struct {
+	v *metrics.Value
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+func (jmv jsonMetricValue) MarshalJSON() ([]byte, error) {
+	switch kind := jmv.v.Kind(); kind {
+	case metrics.KindUint64:
+		return json.Marshal(jmv.v.Uint64())
+	case metrics.KindFloat64:
+		return json.Marshal(jmv.v.Float64())
+	default:
+		return nil, errors.Errorf("can't JSON-encode Go metric value of kind %d", int(kind))
+	}
+}
+
+type SuccessfulSync struct {
+	When time.Time
+	Took time.Duration
+}
+
+// CurrentDbConnErr is to be updated by the DB connector.
+var CurrentDbConnErr com.Atomic[error]
+
+// LastSuccessfulSync is to be updated by the main() function.
+var LastSuccessfulSync com.Atomic[SuccessfulSync]
+
+var boolToStr = map[bool]string{false: "0", true: "1"}
+
+// StartHeartbeat periodically writes heartbeats to Redis for being monitored by Icinga 2.
+func StartHeartbeat(
+	ctx context.Context, client *icingaredis.Client, logger *logging.Logger, ha ha, heartbeat *icingaredis.Heartbeat,
+) {
+	allMetrics := metrics.All()
+	samples := make([]metrics.Sample, 0, len(allMetrics))
+	cumulative := map[string]jsonMetricValue{}
+	notCumulative := map[string]jsonMetricValue{}
+	byCumulative := map[bool]map[string]jsonMetricValue{true: cumulative, false: notCumulative}
+	mtrcs := map[string]map[string]jsonMetricValue{"cumulative": cumulative, "not-cumulative": notCumulative}
```

Comment:
Looks strange to me to use cumulative as the primary property to group metrics by. But let me guess: the idea was to avoid the overhead of serializing lots of `{"value",X,"cumulative":false}`?

Reply:
What about preparing all these metrics in Go? This would save all this code packing them into awkwardly structured data structures. I mean we could simply write a perfdata string to Redis like any other plugin would write it to stdout.